ramips: Add support for Mikrotik RouterBOARD rbm33g

TobleMiner opened the pull request:
This PR adds support for the Mikrotik RouterBOARD rbm33g.

# Hardware

The RBM33g is a mt7621 based device featureing three gigabit ports, 2 miniPCIe slots with simcard sockets, 1 M.2 slot, 1 USB 3.0 port and a male onboard RS-232 serial port. Additionally there are a lot of accessible GPIO ports and additional buses like i2c, mdio, spi and uart.

## Flash

The device has two spi flash chips. The first flash chips is rather small (512 kB), connected to CS0 by default and contains only the RouterBOOT bootloader and some factory information (e.g. mac address).
The second chip has a size of 16 MB, is by default connected to CS1 and contains the firmware image.


# Development Status

Ethernet: ✔️
miniPCIe: ✔️
USB 3.0: ✔️
M.2: ? (Probably working but untested)


# Installation

## Prerequisites

A USB -> RS-232 Adapter and a nullmodem cable are required for installation.

To install an OpenWRT image to the device two components must be built:

1. A factory initramfs image
2. A openwrt sysupgrade image

### initramfs & sysupgrade image

Select target device ```Mikrotik RBM33G``` in openwrt menuconfig and build the images. This will create the images ```openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin``` and ```openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin``` in the output directory.

## Installing

**Make sure to back up your RouterOS license in case you do ever want to go back to RouterOS using ```/system license output``` and back up the created license file.**

Serial settings: 115200 8N1

The installation is a two-step process. First the ```openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin``` must be booted via tftp:

1. Set up a dhcp server that points the bootfile to tftp server serving the ```openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin``` initramfs image
2. Connect to WAN port (left side, next to LEDs and power indicator)
3. Connect to serial port of board
4. Power on board and enter RouterBOOT setup menu
5. Set boot device to ```boot over ethernet```
6. Set boot protocol to ```dhcp protocol``` (can be omitted if DHCP server allows dynamic bootp)
6. Save config
7. Wait for board to boot via Ethernet

On the serial port you should now be presented with the OpenWRT boot log. The next steps will install OpenWRT persistently.

1. Copy ```openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin``` to the device using scp.
2. Write openwrt to flash using ```sysupgrade openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin```

Once the flashing completes reboot the router and let it boot from flash. It should boot straight to OpenWRT.

TobleMiner committed:
ramips: Add support for mt7621 to lzma-loader

Signed-off-by: Tobias Schramm <tobleminer@gmail.com>

TobleMiner committed:
ramips: Add lzma-loader targets

Signed-off-by: Tobias Schramm <tobleminer@gmail.com>

TobleMiner committed:
ramips: Add support for Mikrotik RouterBOARD RBM33g

This commit adds support for the Mikrotik RouterBOARD RBM33g.

=Hardware=

The RBM33g is a mt7621 based device featuring three gigabit ports, 2
miniPCIe slots with sim card sockets, 1 M.2 slot, 1 USB 3.0 port and a male
onboard RS-232 serial port. Additionally there are a lot of accessible
GPIO ports and additional buses like i2c, mdio, spi and uart.

==Switch==

The three Ethernet ports are all connected to the internal switch of the
mt7621 SoC:

port 0: Ethernet Port next to barrel jack with PoE printed on it
port 1: Innermost Ethernet Port on opposite side of RS-232 port
port 2: Outermost Ethernet Port on opposite side of RS-232 port
port 6: CPU

==Flash==

The device has two spi flash chips. The first flash chips is rather small
(512 kB), connected to CS0 by default and contains only the RouterBOOT
bootloader and some factory information (e.g. mac address).
The second chip has a size of 16 MB, is by default connected to CS1 and
contains the firmware image.

==PCIe==

The board features three PCIe-enabled slots. Two of them are miniPCIe
slots (PCIe0, PCIe1) and one is a M.2 (Key M) slot (PCIe2).
Each of the miniPCIe slots is connected to a dedicated mini SIM socket
on the back of the board.

Power to all three PCIe-enabled slots is controlled via GPIOs on the
mt7621 SoC:

PCIe0: GPIO9
PCIe1: GPIO10
PCIe2: GPIO11

==USB==

The board has one external USB 3.0 port at the rear. Additionally PCIe
port 0 has a permanently enabled USB interface. PCIe slot 1 shares its
USB interface with the rear USB port. Thus only either the rear USB port
or the USB interface of PCIe slot 1 can be active at the same time. The
jumper next to the rear USB port controls which one is active:

open: USB on PCIe 1 is active
closed: USB on rear USB port is active

==Power==

The board can accept both, passive PoE and external power via a 2.1 mm
barrel jack. The input voltage range is 11-32 V.

=Installation=

==Prerequisites==

A USB -> RS-232 Adapter and a null modem cable are required for
installation.

To install an OpenWRT image to the device two components must be built:

1. A openwrt initramfs image
2. A openwrt sysupgrade image

===initramfs & sysupgrade image===

Select target devices "Mikrotik RBM33G" in
openwrt menuconfig and build the images. This will create the images
"openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin" and
"openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin" in the output
directory.

==Installing==

**Make sure to back up your RouterOS license in case you do ever want to
go back to RouterOS using "/system license output" and back up the created
license file.**

Serial settings: 115200 8N1

The installation is a two-step process. First the
"openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin" must be booted
via tftp:

1. Set up a dhcp server that points the bootfile to tftp server serving
   the "openwrt-ramips-mt7621-mikrotik_rbm33g-initramfs-kernel.bin"
   initramfs image
2. Connect to WAN port (left side, next to sys-LED and power indicator)
3. Connect to serial port of board
4. Power on board and enter RouterBOOT setup menu
5. Set boot device to "boot over ethernet"
6. Set boot protocol to "dhcp protocol" (can be omitted if DHCP server
   allows dynamic bootp)
6. Save config
7. Wait for board to boot via Ethernet

On the serial port you should now be presented with the OpenWRT boot log.
The next steps will install OpenWRT persistently.

1. Copy "openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin" to the device
   using scp.
2. Write openwrt to flash using "sysupgrade
   openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin"

Once the flashing completes reboot the router and let it boot from flash.
It should boot straight to OpenWRT.

Signed-off-by: Tobias Schramm <tobleminer@gmail.com>

blogic commented:
@TobleMiner looking good. please drop the fullduplex fixup from this patch. i am building a tree just now for a slightly different fix and will push it to trunk when tested. why do we need lzma loader on this target ?

mkresin commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -203,6 +203,36 @@ define Device/rb750gr3
 
+define Device/rbm33g
+  DTS := RBM33G
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
```
your firmware partition is 16515072 bytes but ralink_default_fw_size_16M is 16121856. Please use `IMAGE_SIZE := 16128k` here.

TobleMiner replied:
Yep, of course. I should have adjusted that before pushing

mkresin commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -203,6 +203,36 @@ define Device/rb750gr3
+  IMAGE_SIZE := $(ralink_default_fw_size_16M)
+  DEVICE_TITLE := MikroTik RBM33G
+  BOARDNAME := MIKROTIK-RBM33G
+  DEVICE_PACKAGES := kmod-usb3 uboot-envtools
```
Please keep in mind, that these packages will not be included in the initramfs build by the buildbot. It is a limitation of the CONFIG_TARGET_PER_DEVICE_ROOTFS which is set by the buildbots.

TobleMiner replied:
Thanks, I was not aware of that.

mkresin commented on target/linux/ramips/dts/RBM33G-factory.dts:
```diff
@@ -0,0 +1,170 @@
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		usr {
```
would it make sense to use this LED as boot status indicator (diag.sh)?

mkresin commented on target/linux/ramips/dts/RBM33G-factory.dts:
```diff
@@ -0,0 +1,170 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "uart2", "uart3", "jtag";
```
should be `ralink,group = "uart2", "wdt";`. The wdt pin is used for the reset btn and none of the uart3 pins are used as GPIOs.

mkresin reviewed (changes requested):
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

mkresin commented:
I don't really like the idea of an extra factory image. As far as I can see, the only difference is that the uboot and uboot-env partition are writeable.

I would expect that the lzma loader is used to make the preinstalled bootloader happy, hence no update of the bootloader is required.

The writable u-boot env partition is kind of bogus, as no support for the board is added to package/boot/uboot-envtools/files/ramips.

Nevertheless, I'm fine if both are writeable (if required) for the "normal" image.

Could it be that the GPIOs are connected to the PCIe perstN/reset lines? I've never seen VCC switched via GPIO. The driver has a GPIO_PERST define but the used GPIOs are hardcoded and support for a devicetree property to pass the GPIOs seems to be missing. Using the fixed regulators is the cleanest way so far.

Your commit message lacks an install instruction. Please check the recent board additions for some examples. If possible please extend the hardware description as well. It mentions only the real basics.

TobleMiner commented:
@mkresin 

> I don't really like the idea of an extra factory image. As far as I can see, the only difference is that the uboot and uboot-env partition are writeable.

Almost. They are using two different device trees with swapped chip select lines for the two flash chips. I need that because I have no idea where the flash chip at CS1 is mapped in memory at bootup. Nevertheless I could probably somehow add it as an additional image to the profile.

> I would expect that the lzma loader is used to make the preinstalled bootloader happy, hence no update of the bootloader is required.

Yeah. The preinstalled bootloader can boot elf images over the network and I'm using the LZMA loader to get the initial initramfs image up and running.

> The writable u-boot env partition is kind of bogus, as no support for the board is added to package/boot/uboot-envtools/files/ramips.

It is only writeable in case someone wants to flash a pre prepared image to it. I just figured it would be of no harm.


> Could it be that the GPIOs are connected to the PCIe perstN/reset lines?

Nope, I have verified that the GPIOs are in fact wired to the enable lines of the PCIe voltage regulators.

@blogic 

> please drop the fullduplex fixup 

Can I have a look at the patch? Is it a real fix for duplex mode or does it just disable duplex mode?

> why do we need lzma loader on this target 

The lzma-loader is needed for MIkrotiks RouterBOOT bootloader. It can boot over network but it expects an elf image. Since lzma-loader seems to be the go-to approach for anything requiring elf images I though it would be best to add mt7621 to it.

TobleMiner commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -421,6 +421,9 @@ ramips_board_detect() {
 	*"RB750Gr3")
 		name="rb750gr3"
 		;;
+	*"RBM33G")
```
Ok, I'll drop that. I was not really aware of the fact that the name could be set in that way.

TobleMiner commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -203,6 +203,36 @@ define Device/rb750gr3
 endef
 TARGET_DEVICES += rb750gr3
 
+define Device/rbm33g
```
Ok, cool. Will do so

TobleMiner commented:
@blogic 

I've removed the patch disabling full duplex spi transfers.

@mkresin 

I've implemented your suggestions/fixes. A lot of the stuff was not obvious to me while developing. Thanks for pointing it out to me.

There are still two targets, one for factory and one for upgrading. I'm not really sure how to get rid of the factory image. I think I'd need to have more control over creation of initramfs images to eliminate the factory target. I think something like a ```append-initramfs```command for the image creation "pipeline" would be quite useful.

blogic commented on target/linux/ramips/base-files/etc/diag.sh:
```diff
@@ -157,6 +157,9 @@ get_status_led() {
 	zorlik,zl5900v2)
 		status_led="$boardname:blue:power"
 		;;
+	mikrotik,rbm33g)
```
alphabetical ordering please

TobleMiner replied:
Whoops, didn't quite catch the ordering there, thanks

blogic commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -197,6 +197,7 @@ ramips_setup_interfaces()
 	jhr-n926r|\
 	mzk-wdpr|\
 	rb750gr3|\
+	mikrotik,rbm33g|\
```
alphabetical ordering please

mkresin commented on target/linux/ramips/mt7621/config-4.14:
```diff
@@ -284,3 +284,7 @@ CONFIG_WEAK_ORDERING=y
 CONFIG_ZLIB_INFLATE=y
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_GPIO=y
+CONFIG_REGULATOR_FIXED=y
```
This config symbol doesn't exists. Do you mean `CONFIG_REGULATOR_FIXED_VOLTAGE` by any chance? Furthermore you broke alphabetical order here. Please consider using `make kernel_menuconfig CONFIG_TARGET=subtarget` to add kernel config symbols.

I'm quite sure  `REGULATOR_GPIO` isn't required. It's meant for regulators which set different voltage levels based on the GPIO (input) value.

Am I right that the PCIe vcc is enabled by default by the bootloader? At least with this code, OpenWrt never enabled them.

TobleMiner replied:
Uh, thanks for pointing that out. Yes, the bootloader enables the PCIe power GPIOs because mt7621-pci driver lacks hotplug support. Slots where no device is detected on bootup end up permanently disabled.

mkresin commented:
> There are still two targets, one for factory and one for upgrading. I'm not really sure how to get rid of the factory image. I think I'd need to have more control over creation of initramfs images to eliminate the factory target. 

You have using `KERNEL_INITRAMFS` in your image recipe. By default it is `KERNEL_INITRAMFS = $$(KERNEL)`. Grep for `KERNEL_INITRAMFS` to find some examples.

TobleMiner commented:
@mkresin Thanks. That's almost what I was searching for. Using this method will however break the normal initramfs image for booting via uboot after flashing because 
1. The rather old uboot I'm using can't boot elf images
2. The chip select pins would be swapped

Thus I was searching for a method to create additional initramfs images.

zeilsteen2 commented:
Thanks for the great work! I have been looking for this for a while. Great you got it working.

Noob question:
I can build images with Lede but how do I add this updated device to the toolchain? I got the latest checked in but I don't see the device yet with make menuconfig. Thank you.

TobleMiner commented:
@zeilsteen2 You need to freshly clone my openwrt repo:
```
git clone -b feature-rbm33g-rebase https://github.com/TobleMiner/openwrt
```

That will give you a openwrt source tree with the rbm33g added. After that you can do all the usual feeds update/install and menuconfig stuff in there and you should see the ```MikroTik RBM33G``` as a target.

TobleMiner commented:
@mkresin Well, the problem is that the bootloader must be on the flash attached to cs0. That's a hardware limitation. But since I do only know how to make the factory bootloader, RouterBOOT, work with images booted via network but not with ones stored on flash I have to replace it with u-boot.
But again in the u-boot source the code for handling flash chips attached to CS1 is missing. Thus the chip containing the kernel must be on CS0.

Thus the factory dtb must have the big flash chip on CS1 since RouterBOOT is stored on the small one and that must be on CS0.
But when booting OpenWRT from flash we have to boot from the big flash chip since OpenWRT does not fit on the small flash chip. Thus it has to be attached to CS0 now.

Additionally the flash layout of the small flash chip does not really allow installing u-boot there. U-boot is quite a bit bigger than the factory RouterBOOT bootloader and the factory config partition starts just after RouterBOOT and holds important values like the devices mac addresses. Thus I can not really store u-boot on the smaller flash chip.

A possible workaround might be:

- small flash chip connected to CS0, large one to CS1
- boot board and enter RouterBOOT cli
- swap chip selects
- boot initramfs image

That would keep the chip select mapping OpenWRT sees consistent.

But I fear that any flashing procedure that involves complex sequences of actions while the hardware is powered on might deter people from trying openwrt on this device and make flashing unnecessarily hard.

TobleMiner commented:
@zeilsteen2

I forgot one important thing: You will need to apply another patch on the source tree to make it work.

Unzip [rbm33g_spi_patch.zip](https://github.com/openwrt/openwrt/files/2116018/rbm33g_spi_patch.zip)
and copy the file rbm33g_spi.patch to your openwrt source tree, then run ```git apply  rbm33g_spi.patch``` before building. It's a workaround for prolems with full duplex spi data transfer on the MT7621A

TobleMiner commented:
@mkresin  Nope, haven't tried that yet. I'll have to give it a go.

Since you are already here, would you mind taking a look at #689 ? This PR has been sitting there for quite a while although a lot of people have reported great success.

zeilsteen2 commented:
@TobleMiner  thanks for the help, I cloned your version of openwrt and did the updates from the readme. But I don't see the target yet, also after the patch only the old Mikrotik entry as target.  When I do a menuconfig I get some dependency errors over libpam,libgnutls,libopenldap,libidn2,liblzma,libnetsnmp.

Would that be the cause? I also only have RBM33G.dts and RBM33G-factory.dts if I do a search in the dir.

Thanks.

TobleMiner commented:
@zeilsteen2 THis works for me:

```
[tsys@buidler openwrt]$ git clone -b feature-rbm33g-rebase https://github.com/TobleMiner/openwrt rbm33g-verify
[tsys@buidler openwrt]$ cd rbm33g-verify/
[tsys@buidler rbm33g-verify]$ ./scripts/feeds update -a
[tsys@buidler rbm33g-verify]$ ./scripts/feeds install -a
[tsys@buidler rbm33g-verify]$ cp ../rbm33g_spi.patch .
[tsys@buidler rbm33g-verify]$ git apply rbm33g_spi.patch 
[tsys@buidler rbm33g-verify]$ make menuconfig
```
Inside menuconfig I select the targte and the subtarget
![menuconfig](https://user-images.githubusercontent.com/2042858/41612787-35030732-73f4-11e8-8694-e98879a8c573.png)

As you can see I can select the Mikrotik RBM33G just fine.

mkresin commented:
I'm sorry, ar71xx isn't my party.

TobleMiner commented:
@mkresin No problem. Any ideas on who I could ask to take a look at it?

zeilsteen2 commented:
@TobleMiner  you're a genius, thanks for pointing out the profile! I will be building in a minute 👍 
I also flashed u-boot directly like with the RB750Gr3 to the winbond flash using spi and changed the jumpers. That is booting and now kermitting another test kernel to see if it runs.  And after that will try your buildchain. :))

mkresin commented:
I already mentioned f00b4r0 who added some ar71xx mikrotik boards.

Please have a look at https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ar71xx/base-files/lib/upgrade/platform.sh;h=9bc3f8564b12dac03f5eb0d0b9ad71cc072c56ca;hb=HEAD#l717 as well. As far as I remember, the usual mikrotik installation is done via initramfs image. During installation from the initramfs image, the existing firmware partition needs to be completely erased, to ensure the bootloader picks our kernel and not some leftover which looks like our kernel. Or something like that.

Unfortunately I can't provide more details, as it is all I remember to have heard/read. I never touched mikrotik hardware on my own.

zeilsteen2 commented:
@TobleMiner, the kernel is running from tftp, fantastic! Do you know if it possible to add wil6210 module in the build? You can't select it from the menuconfig but in other versions you can.  I added it by hand in the .config file but that is not adding it I think.  I want to check if the cards run on the board. (cards are pci-e 2.0 and I think this board had pci-e 1.1 but I hope the cards wil just take the lower clockspeed and run. The pci-e ID is detected so there is some action.

TobleMiner commented:
@zeilsteen2 I've never seen that option in OpenWrt menuconfig before. Also I have not really done anything with 60 GHz wireless beyond using some MikroTik wireless wires yet. I don't think I'll be of much use regarding drivers for 60 GHz wireless.

zeilsteen2 commented:
@TobleMiner  that's indeed the card i'm looking for, but somehow my view in the menuconfig is not showing it.  But like you pointed out I need to run make kernel_menuconfig CONFIG_TARGET=mt7621 than. 
Thanks for showing.

mkresin commented:
@TobleMiner please remove "define Build/loader-kernel-cmdline". It will be never required for ramips and is rather a copy/paste leftover from the not devicetree target ar71xx.

Please add the [platform_pre_upgrade()](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ar71xx/base-files/lib/upgrade/platform.sh;h=9bc3f8564b12dac03f5eb0d0b9ad71cc072c56ca;hb=HEAD#l717)  for your board. Please refer to the sysupgrade command instead of 'mtd write' in your commit message. This way you ensure that there are no mikrotik kernel leftovers on first boot after flash, which might cause issues as it was observed with ar71xx/mikrotik in the past.

I'll try to find a user of the DIR-860L B1, which uses the relocate part of the lzma loader. I would like to ensure that your changes don't cause a regression for this board.

TobleMiner commented:
@mkresin 

> please remove "define Build/loader-kernel-cmdline". It will be never required for ramips and is rather a copy/paste leftover from the not devicetree target ar71xx.

Yep, indeed. That's a leftover copying the targets from ar71xx.

> Please add the platform_pre_upgrade() for your board. Please refer to the sysupgrade command instead of 'mtd write' in your commit message.

Oops, sorry. Totally skipped that part on your last comment. That actually makes a lot of sense. When looking at a factory flash image of a rbm33g I can see multiple yaffs filesystems in different locations on flash.

TobleMiner commented:
@mkresin I've implemented the changes you have suggested.

> I'll try to find a user of the DIR-860L B1, which uses the relocate part of the lzma loader. I would like to ensure that your changes don't cause a regression for this board.

Good catch, I was under the impression that no target at all was using lzma-loader before. But I didn't realize it was part of relocate-kernel. It is certainly a good idea to find someone with a DIR-860L B1 to ensure i didn't break it.

mweinelt commented:
I sysupgraded to 18.06 + your changes on my DIR-860L-B1 and it rebooted just fine. Anything else that I would need to test?

TobleMiner commented:
@mweinelt I think that should be sufficient. Thanks a bunch for testing!

mkresin commented:
merged, thanks!

m7mdcc commented:
Hi , Thanks for support ! i have tried your steps but after flashing openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-sysupgrade.bin to device , its say "Could not open mtd device: firmware" .. how i can fix this ?

this help ?
cat /proc/mtd
dev: size erasesize name
mtd0: 0000f000 00010000 "routerboot"
mtd1: 00071000 00010000 "factory"

TobleMiner commented:
Hm, @blogic what's up with the spi patch? Is there no upstream patch to make spi0.1 work?

@m7mdcc Take a look at this https://github.com/openwrt/openwrt/pull/925#issuecomment-398459513 There is an additional patch you might have to apply. Seems like there is no upstream fix for it although @blogic asked me to remove my patch from this PR.

@mkresin Maybe we should actually merge my patch for spi0.1?

m7mdcc commented:
@TobleMiner  , I think your repo was already patched?  this one : git clone -b feature-rbm33g-rebase https://github.com/TobleMiner/openwrt , right?

TobleMiner commented:
@m7mdcc Nope, that's the source for this repo and is thus not patched. You'll have to apply the patch on top of my support branch.

m7mdcc commented:
@TobleMiner  I tried your support branch with applying the patch its same behavior . 
did i miss something else ? 
`root@OpenWrt:/tmp# sysupgrade -n openwrt-ramips-mt7621-mikrotik_rbm33g-squashfs-
sysupgrade.bin
Commencing upgrade. All shell sessions will be closed now.
Watchdog handover: fd=3
- watchdog -
killall: telnetd: no process killed
Sending TERM to remaining processes ... ntpd dnsmasq ubusd logd netifd odhcpd
Sending KILL to remaining processes ...
Could not open mtd device: firmware
Could not open mtd device: firmware
Performing system upgrade...
Could not open mtd device: firmware
Can't open device for writing!
Upgrade completed
Rebooting system...
umount: can't unmount /dev: Resource busy
umount: can't unmount /tmp: Resource busy
umount: can't unmount /: Invalid argument
[  151.673655] reboot: Restarting system
`

m7mdcc commented:
this help ? 
[    3.401402] spi-mt7621 1e000b00.spi: sys_freq: 50000000
[    3.417918] m25p80 spi0.0: w25x40 (512 Kbytes)
[    3.422387] 2 fixed-partitions partitions found on MTD device spi0.0
[    3.428725] Creating 2 MTD partitions on "spi0.0":
[    3.433502] 0x000000000000-0x00000000f000 : "routerboot"
[    3.439713] 0x00000000f000-0x000000080000 : "factory"
[    3.446015] m25p80 spi0.1: unrecognized JEDEC id bytes: ef, 70, 18
[    3.452206] m25p80: probe of spi0.1 failed with error -2
[    3.458041] libphy: Fixed MDIO Bus: probed

TobleMiner commented:
@m7mdcc Well, that's interesting. That's indeed an unsupported SPI flash. I've only seen W25Q128FV on the rbm33g yet. #995 seem very related. We'll need support for the JV series first.

m7mdcc commented:
@TobleMiner i have fix my issue by patch on #995 .

Thanks for your support !

TobleMiner commented:
@m7mdcc Nice. Btw thanks for the info about JV winbond flash on MikroTik boards :smiley:
